contracts/NestedFactory.sol

```diff
@@ -181,117 +163,134 @@ contract NestedFactory is INestedFactory, ReentrancyGuard, OwnableProxyDelegatio
         }
 
         emit NftCreated(nftId, _originalTokenId);
     }
 
     /// @inheritdoc INestedFactory
-    function processInputOrders(uint256 _nftId, BatchedInputOrders[] calldata _batchedOrders)
-        external
-        payable
-        override
-        nonReentrant
-        onlyTokenOwner(_nftId)
-        isUnlocked(_nftId)
-    {
-        _checkMsgValue(_batchedOrders);
+    function processInputOrders(
+        uint256 _nftId,
+        address _feesToken,
+        uint256 _feesAmount,
+        BatchedInputOrders[] calldata _batchedOrders
+    ) external payable override nonReentrant onlyTokenOwner(_nftId) isUnlocked(_nftId) {
+        _checkMsgValue(_batchedOrders, _feesToken, _feesAmount);
+
+        // Pay fees
+        _transferInputTokens(_nftId, IERC20(_feesToken), _feesAmount, false);
+        _transferFees(_feesToken, _feesAmount, _nftId);
+
         _processInputOrders(_nftId, _batchedOrders);
         emit NftUpdated(_nftId);
     }
 
     /// @inheritdoc INestedFactory
-    function processOutputOrders(uint256 _nftId, BatchedOutputOrders[] calldata _batchedOrders)
-        external
-        override
-        nonReentrant
-        onlyTokenOwner(_nftId)
-        isUnlocked(_nftId)
-    {
+    function processOutputOrders(
+        uint256 _nftId,
+        address _feesToken,
+        uint256 _feesAmount,
+        BatchedOutputOrders[] calldata _batchedOrders
+    ) external override nonReentrant onlyTokenOwner(_nftId) isUnlocked(_nftId) {
+        // Pay fees
+        _transferInputTokens(_nftId, IERC20(_feesToken), _feesAmount, false);
+        _transferFees(_feesToken, _feesAmount, _nftId);
+
         _processOutputOrders(_nftId, _batchedOrders);
         emit NftUpdated(_nftId);
     }
 
     /// @inheritdoc INestedFactory
     function processInputAndOutputOrders(
         uint256 _nftId,
+        address _feesToken,
+        uint256 _feesAmount,
         BatchedInputOrders[] calldata _batchedInputOrders,
         BatchedOutputOrders[] calldata _batchedOutputOrders
     ) external payable override nonReentrant onlyTokenOwner(_nftId) isUnlocked(_nftId) {
-        _checkMsgValue(_batchedInputOrders);
+        _checkMsgValue(_batchedInputOrders, _feesToken, _feesAmount);
+
         _processInputOrders(_nftId, _batchedInputOrders);
         _processOutputOrders(_nftId, _batchedOutputOrders);
+         // Pay fees
+        _transferInputTokens(_nftId, IERC20(_feesToken), _feesAmount, false);
+        _transferFees(_feesToken, _feesAmount, _nftId);
+
         emit NftUpdated(_nftId);
     }
 
     /// @inheritdoc INestedFactory
     function destroy(
         uint256 _nftId,
+        address _feesToken,
+        uint256 _feesAmount,
         IERC20 _buyToken,
         Order[] calldata _orders
     ) external override nonReentrant onlyTokenOwner(_nftId) isUnlocked(_nftId) {
         address[] memory tokens = nestedRecords.getAssetTokens(_nftId);
-        uint256 tokensLength = tokens.length;
```

Comment:
Did removing the `tokens.length` cache reduce the gas cost?

`tokens.length` is now called 2 times, in lines 251 and 260.

Reply:
Unfortunately, we face a stack too deep when we add back this variable.
It seems we cannot avoid this with the new parameters